flatland/envs/agent_chains.py

```diff
@@ -270,205 +248,136 @@ def check_motion(self, iAgent, rcPos):
+        """
+        if r1 is None:
+            r1 = (None, i)
+        if r2 is None:
+            r2 = (None, i)
+        self.agents[i] = (r1, r2)
+        if r2 not in self.reverse_target:
+            self.reverse_target[r2] = [i]
+        else:
+            self.reverse_target[r2].append(i)
 
-    cte.nextRow()
+    def find_conflicts(self):
+        """
+        Find and resolve conflicts:
+        - swaps aka. deadlocks (head-to-head collisions)
+        - two agents same target
 
-    cte.addAgentToRow(1, 2, "Land on\nSame")
-    cte.addAgentToRow(3, 2)
+        Correctness:
+        - deadlocked agents are stopped
+        - for each conflict, one of the agents is stopped.
 
-    cte.nextRow()
-    cte.addAgentToRow(1, 2, "chains\nonto\nsame")
-    cte.addAgentToRow(2, 3)
-    cte.addAgentToRow(3, 4)
-    cte.addAgentToRow(5, 4)
-    cte.addAgentToRow(6, 5)
-    cte.addAgentToRow(7, 6)
+        Termination: The list of target_conflicts will eventually be empty as in every round, one agent is stopped.
 
-    cte.nextRow()
-    cte.addAgentToRow(1, 2, "3-way\nsame")
-    cte.addAgentToRow(3, 2)
-    cte.addAgent((cte.iRowNext + 1, 2), (cte.iRowNext, 2))
-    cte.nextRow()
+        """
+        target_conflicts = self._construct_graph()
+
+        self._fix_conflicts(target_conflicts)
+
+    def _construct_graph(self):
+        target_conflicts: List[Tuple[AgentHandle, AgentHandle]] = []  # a1 < a2
+        for iAg, (pos, target) in self.agents.items():
+            conflict_list = []
+            if pos in self.reverse_target:
+                conflict_list += self.reverse_target[pos]
+            if target in self.reverse_target:
+                conflict_list += self.reverse_target[target]
+            for a2 in conflict_list:
+                if iAg >= a2:
+                    continue
+                a2pos, a2_target = self.agents[a2]
+                # same target
+                if a2_target == target:
+                    target_conflicts.append((iAg, a2))
+                # swap/collision
+                elif pos == a2_target and target == a2pos:
+                    self.stopped.add(iAg)
+                    self.stopped.add(a2)
+                    self.deadlocked.add(iAg)
+                    self.deadlocked.add(a2)
+            if pos == target:
+                self.stopped.add(iAg)
+        # no need to process deadlocked agents further (avoid ConcurrentModificationException)
+        for a in self.deadlocked:
+            self.agents[a] = (self.agents[a][0], self.agents[a][0])
+        return target_conflicts
+
+    def _fix_conflicts(self, target_conflicts):
+        while len(target_conflicts) > 0:
+            u, v = target_conflicts[0]
+
+            u_pos, u_target = self.agents[u]
+            v_pos, v_target = self.agents[v]
+            if u in self.stopped and v in self.stopped:
+                target_conflicts = target_conflicts[1:]
+                continue
```

Comment:
Not needed in paper as looping over edges there. Double check this is equivalent.